### PR TITLE
Add check for the correct layout of camera intrinsics matrices.

### LINKED ISSUE #342: No check in place if camera intrinsics matrices have the correct layout
Neither the schema files for protobuf/flatbuffers nor the server check if the following transferred matrices have the right layout:

- distortion matrix
- intrinsic matrix
- rectification matrix
- projection matrix

This for example causes the server to crash randomly on my machine, when running the script [gRPC_pb_sendLabeledImageGrid.py](https://github.com/agri-gaia/seerep/blob/bug/camintrins-loading-projects/examples/python/gRPC/images/gRPC_pb_sendLabeledImageGrid.py) beforehand, then moving the created project out of the folder, restarting the server, moving the project file back and finally executing the script [gRPC_fb_loadProjects](https://github.com/agri-gaia/seerep/blob/bug/camintrins-loading-projects/examples/python/gRPC/meta/gRPC_fb_loadProjects.py). 
Note: these steps were done on the **bug/camintrins-loading-projects** branch.
The intrics matrix in the example only has three values.

The problem lies in here where the camera intrics matrix is used to calculate the frustum:
https://github.com/agri-gaia/seerep/blob/ea2f60522b305c271ba6056de0d790c1c9d035de/seerep_hdf5/seerep_hdf5_core/src/hdf5_core_image.cpp#L138-L139

Here the 5th value of the intrisics matrix is accessed, but not set by the example script when transferring the images. This causes undefined behaviour and the crash of the server after loading some images. The debugger shows that in this case eventually one image has it's `field_of_view_y` set to `inf`.

Possible solutions which came to my mind researching this problem could be:
1. set fixed sizes for the matrices in the schema files using flatbuffers structs and protobufs nested messages.
2. checking for the correct layout on the server side.

### PR DESCRIPTION
**Pull request by mmeijerdfki**
Closes #342. This also corrects a typo in Frustum.

Edit:
Empty values are now allowed, except for the intrinsics matrix.

### CODE REVIEW

File: seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp [hunk 1]
```diff
@@ -68,6 +68,48 @@ grpc::Status FbCameraIntrinsicsService::TransferCameraIntrinsics(
...
+  bool distortion_model_err =
+      !requestRoot->distortion_model() ||
+      std::find(std::begin(seerep_server_constants::kCameraDistortionModels),
+                std::end(seerep_server_constants::kCameraDistortionModels),
+                requestRoot->distortion_model()->c_str()) == std::end(seerep_server_constants::kCameraDistortionModels);
+
+  bool distortion_mat_err = !requestRoot->distortion() || requestRoot->distortion()->size() < 1;
+  bool intrinsics_mat_err = !requestRoot->intrinsic_matrix() || requestRoot->intrinsic_matrix()->size() != 9;
+  bool rectification_mat_err = !requestRoot->rectification_matrix() || requestRoot->rectification_matrix()->size() != 9;
+  bool projection_mat_err = !requestRoot->projection_matrix() || requestRoot->projection_matrix()->size() != 12;
```
Comment (Mark-Niemeyer) [inline seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp:81]: only the intrinsic mat is a must have. The rest may be empty if the information is unknown. The check should allow empty mat for these.

File: seerep_srv/seerep_server/src/pb_camera_intrinsics_service.cpp [hunk 2]
```diff
@@ -14,6 +14,47 @@ grpc::Status PbCameraIntrinsicsService::TransferCameraIntrinsics(grpc::ServerCon
...
+
+  // check if the distortion model is found in DISTORTION_MODELS
+  bool distortion_model_err = std::find(std::begin(seerep_server_constants::kCameraDistortionModels),
+                                        std::end(seerep_server_constants::kCameraDistortionModels),
+                                        camintrinsics->distortion_model().c_str()) ==
+                              std::end(seerep_server_constants::kCameraDistortionModels);
+  bool distortion_mat_err = camintrinsics->distortion_size() < 1;
+  bool intrinsics_mat_err = camintrinsics->intrinsic_matrix_size() != 9;
+  bool rectification_mat_err = camintrinsics->rectification_matrix_size() != 9;
+  bool projection_mat_err = camintrinsics->projection_matrix_size() != 12;
```
Comment (Mark-Niemeyer) [inline seerep_srv/seerep_server/src/pb_camera_intrinsics_service.cpp:27]: see fb

File: seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp [hunk 3]
```diff
@@ -68,6 +68,48 @@ grpc::Status FbCameraIntrinsicsService::TransferCameraIntrinsics(
...
+
+  if (distortion_model_err || distortion_mat_err || intrinsics_mat_err || rectification_mat_err || projection_mat_err)
+  {
+    if (distortion_model_err)
+    {
+      response_message = "The distortion model is not set or invalid.";
+    }
+    else if (distortion_mat_err)
+    {
+      response_message = "The distortion matrix has to be set correctly in relation to the distortion model.";
```
Comment (Mark-Niemeyer) [inline seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp:91]: maybe use += and no "else if" to return all errors at once?

**Review by Mark-Niemeyer — CHANGES_REQUESTED** [review]

File: seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp (hunk 1 above)
Comment (mmeijerdfki) [reply seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp:81]: Only allowing empty matrices causes another error while writing the values to hdf5 for pb at least. Maybe checking that a matrix has only one entry could signal that its values should be ignored.

File: seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp [hunk 4]
```diff
@@ -68,6 +68,48 @@ grpc::Status FbCameraIntrinsicsService::TransferCameraIntrinsics(
...
+      response_message += "Rectification matrix dimensions are incorrect. ";
+    }
+    if (projection_mat_err)
+    {
+      response_message += "Projection matrix dimensions are incorrect.";
+    }
+    BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::error) << response_message;
+    seerep_server_util::createResponseFb(response_message, seerep::fb::TRANSMISSION_STATE_FAILURE, response);
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, response_message);
+  }
```
Comment (jarkenau) [inline seerep_srv/seerep_server/src/fb_cameraintrinsics_service.cpp:111]: The checks for Protocol Buffers and Flatbuffers look very similar and could be combined. However, we currently don't have a common place to share this implementation. I think it's fine for now, but we should keep this in mind.

**Review by jarkenau — APPROVED** [review]

**Review by Mark-Niemeyer — APPROVED** [review]

**Pull request merged**